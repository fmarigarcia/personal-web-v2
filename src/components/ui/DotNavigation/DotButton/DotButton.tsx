import React from 'react';
import clsx from 'clsx';

interface DotButtonProps {
    isActive: boolean;
    label: string;
    onClick: () => void;
}

export const DotButton: React.FC<DotButtonProps> = ({
    isActive,
    label,
    onClick,
}) => {
    return (
        <button
            onClick={onClick}
            aria-label={`Go to ${label} section`}
            className={clsx(
                'group relative block w-3 h-3 rounded-full border-2',
                'transition-all duration-300 ease-in-out',
                'hover:scale-125 focus:outline-none focus:ring-2',
                'focus:ring-stone-400 focus:ring-offset-2',
                {
                    'bg-stone-950 border-stone-950 scale-110': isActive,
                    'bg-transparent border-stone-400 hover:border-stone-950':
                        !isActive,
                }
            )}
            type="button"
        >
            {/* Tooltip */}
            <span
                className={clsx(
                    'absolute right-full mr-3 px-2 py-1',
                    'bg-stone-800 text-white text-xs rounded',
                    'opacity-0 pointer-events-none',
                    'transition-opacity duration-200',
                    'whitespace-nowrap top-1/2 -translate-y-1/2',
                    'group-hover:opacity-100'
                )}
            >
                {label}
            </span>
        </button>
    );
};
