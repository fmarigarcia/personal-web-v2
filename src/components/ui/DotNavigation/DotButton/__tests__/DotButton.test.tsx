import { render, screen, fireEvent } from '@testing-library/react';
import { DotButton } from '../DotButton';

describe('DotButton', () => {
    const defaultProps = {
        isActive: false,
        label: 'Home',
        onClick: jest.fn(),
    };

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should render button with correct aria-label', () => {
        render(<DotButton {...defaultProps} />);

        const button = screen.getByRole('button');
        expect(button).toBeInTheDocument();
        expect(button).toHaveAttribute('aria-label', 'Go to Home section');
    });

    it('should render tooltip with label text', () => {
        render(<DotButton {...defaultProps} />);

        const tooltip = screen.getByText('Home');
        expect(tooltip).toBeInTheDocument();
    });

    it('should call onClick when clicked', () => {
        render(<DotButton {...defaultProps} />);

        const button = screen.getByRole('button');
        fireEvent.click(button);

        expect(defaultProps.onClick).toHaveBeenCalledTimes(1);
    });

    it('should apply active styles when isActive is true', () => {
        render(<DotButton {...defaultProps} isActive={true} />);

        const button = screen.getByRole('button');
        expect(button).toHaveClass(
            'bg-blue-600',
            'border-blue-600',
            'scale-110'
        );
    });

    it('should apply inactive styles when isActive is false', () => {
        render(<DotButton {...defaultProps} isActive={false} />);

        const button = screen.getByRole('button');
        expect(button).toHaveClass('bg-transparent', 'border-gray-400');
        expect(button).not.toHaveClass(
            'bg-blue-600',
            'border-blue-600',
            'scale-110'
        );
    });

    it('should have proper button type', () => {
        render(<DotButton {...defaultProps} />);

        const button = screen.getByRole('button');
        expect(button).toHaveAttribute('type', 'button');
    });

    it('should have proper accessibility classes', () => {
        render(<DotButton {...defaultProps} />);

        const button = screen.getByRole('button');
        expect(button).toHaveClass(
            'focus:outline-none',
            'focus:ring-2',
            'focus:ring-blue-500',
            'focus:ring-offset-2'
        );
    });

    it('should have transition and hover classes', () => {
        render(<DotButton {...defaultProps} />);

        const button = screen.getByRole('button');
        expect(button).toHaveClass(
            'transition-all',
            'duration-300',
            'ease-in-out',
            'hover:scale-125'
        );
    });

    it('should have group class for tooltip hover effect', () => {
        render(<DotButton {...defaultProps} />);

        const button = screen.getByRole('button');
        expect(button).toHaveClass('group');
    });

    it('should render tooltip with proper positioning classes', () => {
        render(<DotButton {...defaultProps} />);

        const tooltip = screen.getByText('Home');
        expect(tooltip).toHaveClass(
            'absolute',
            'right-full',
            'mr-3',
            'top-1/2',
            '-translate-y-1/2',
            'opacity-0',
            'group-hover:opacity-100'
        );
    });

    it('should match snapshot when inactive', () => {
        const { container } = render(<DotButton {...defaultProps} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot when active', () => {
        const { container } = render(
            <DotButton {...defaultProps} isActive={true} />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
