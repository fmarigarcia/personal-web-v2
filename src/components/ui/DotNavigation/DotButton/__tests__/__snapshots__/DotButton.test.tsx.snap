// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DotButton should match snapshot when active 1`] = `
<button
  aria-label="Go to Home section"
  class="group relative block w-3 h-3 rounded-full border-2 transition-all duration-300 ease-in-out hover:scale-125 focus:outline-none focus:ring-2 focus:ring-stone-400 focus:ring-offset-2 dark:focus:ring-green-400 dark:focus:ring-offset-zinc-950 bg-stone-950 border-stone-950 scale-110 dark:bg-green-400 dark:border-green-400 dark:shadow-lg dark:shadow-green-400/20"
  type="button"
>
  <span
    class="absolute right-full mr-3 px-2 py-1 bg-stone-800 text-white text-xs rounded dark:bg-zinc-900 dark:text-green-200 dark:border dark:border-zinc-700 opacity-0 pointer-events-none transition-opacity duration-200 whitespace-nowrap top-1/2 -translate-y-1/2 group-hover:opacity-100"
  >
    Home
  </span>
</button>
`;

exports[`DotButton should match snapshot when inactive 1`] = `
<button
  aria-label="Go to Home section"
  class="group relative block w-3 h-3 rounded-full border-2 transition-all duration-300 ease-in-out hover:scale-125 focus:outline-none focus:ring-2 focus:ring-stone-400 focus:ring-offset-2 dark:focus:ring-green-400 dark:focus:ring-offset-zinc-950 bg-transparent border-stone-400 hover:border-stone-950 dark:border-zinc-600 dark:hover:border-green-400"
  type="button"
>
  <span
    class="absolute right-full mr-3 px-2 py-1 bg-stone-800 text-white text-xs rounded dark:bg-zinc-900 dark:text-green-200 dark:border dark:border-zinc-700 opacity-0 pointer-events-none transition-opacity duration-200 whitespace-nowrap top-1/2 -translate-y-1/2 group-hover:opacity-100"
  >
    Home
  </span>
</button>
`;
